List sources and Get source using Sources Client

adiabramovitch opened the pull request:
Depends on https://github.com/RHEnVision/provisioning-backend/pull/31

adiabramovitch commented on internal/routes/all_routes.go:
```diff
@@ -39,5 +39,12 @@ func apiRouter() http.Handler {
 		})
 	})
 
+	r.Route("/sources", func(r chi.Router) {
+		r.Get("/", s.ListSources)
+		r.Route("/{ID}", func(r chi.Router) {
+			r.Get("/", s.GetSource)
+		})
+	})
+
```
Need to decide if sources/aws or sources?provider=aws

adiabramovitch replied:
yes I agree

lzap commented:
I've created a PR which slightly modifies the Makefile so tell me what you think: https://github.com/RHEnVision/provisioning-backend/pull/39

I see a pattern for configuration: for each REST API we will need a configurable option, so perhaps it is worth creating a common configuration group:

```yaml
restEndpoints:
  imageBuilder:
    url: xxxx
  sources:
    url: xxxx
    otherOption: xxxx
```

We can only export RestEndpoints struct, no need to export all values.

adiabramovitch commented:
> 

Sounds good to me, we can always change it if ou would need more information about the client.

lzap commented:
There are conflicts FYI.

lzap commented on internal/payloads/sources_payload.go:
```diff
@@ -0,0 +1,32 @@
+
+type ShowSourceResponse struct {
+	sources.Source
+}
```
Also I would use pointer in here, generally speaking not using pointers is only good for small structs as a very tight optimalization technique.

adiabramovitch commented on internal/db/seeds/dev_small.sql:
```diff
@@ -5,7 +5,7 @@
 
 INSERT INTO accounts(id, account_number, org_id)
-VALUES (1, '13', '000013'),
+VALUES (1, '6089719', '000013'),
```
I have changed it in order to authenticate both sources and our service :)

ezr-ondrej commented:
Are we adding tests, or shall that be a follow-up (which I'm totally fine with)? :)

lzap commented on internal/clients/sources/sources_client.gen.go:
```diff
@@ -1,7 +1,7 @@
+package sources
 
 import (
 	"bytes"
```
Please do not make manual edits to generate files. Remove it from this patch, I will do a PR to fix this. It looks like generation does not work correctly even.

lzap replied:
FYI the PR is here, WIP: https://github.com/RHEnVision/provisioning-backend/pull/46

lzap reviewed (changes requested):
Lookin' good!

adiabramovitch commented on internal/services/sources_service.go:
```diff
@@ -0,0 +1,61 @@
+
+	if err := render.RenderList(w, r, payloads.NewListSourcesResponse(resp.JSON200.Data)); err != nil {
+		// TODO: use the errors supported by the JSON 400 OR 404
+		renderError(w, r, payloads.NewRenderError(r.Context(), "list sources", err))
```
No, In the resp struct there are also errors documented which I want to show as well on a different PR.

adiabramovitch commented on internal/services/sources_service.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+func GetSource(w http.ResponseWriter, r *http.Request) {
+	id := chi.URLParam(r, "ID")
```
I want it as a string not an int

lzap replied:
Sorry disregard.

adiabramovitch commented:
@lzap , @ezr-ondrej I will open today PRs for adding more error handling and the UI dropdown. 
Regarding tests, I'll add those as well in a different PR I want the PRs to be as small as possible. 

@lzap  I addressed your comments.

ezr-ondrej reviewed (commented):
I do not know how to test a happy path (how do I seed sources data?), but when sources return 404, we panic...

adiabramovitch commented on internal/db/seeds/dev_small.sql:
```diff
@@ -8,7 +8,8 @@ INSERT INTO accounts(id, account_number, org_id)
        (3, NULL, '000042'),
-       (4, NULL, '000077')
+       (4, NULL, '000077'),
+       (5, '6089719', '000016')
```
You need to make sure you have this org_id linked to this account number in sources. 
(It is there on default) @ezr-ondrej

lzap replied:
Sure makes sense, but not all our endpoints will interact with sources. Our own accounts still make sense.

lzap commented on internal/services/sources_service.go:
```diff
@@ -0,0 +1,53 @@
+	}
+	resp, err := client.ListSourcesWithResponse(r.Context(), &sources.ListSourcesParams{}, AddIdentityHeader)
+	if err != nil {
+		// TODO: use the errors supported by the JSON 400 OR 404
```
What does this mean? Obviously we want to use 404 when a resource is not found, which is probably not the case for listing operation.

The code you have is good, you return 3rd party error when unknown error occurs which is HTTP 500 which is correct I think. That is the code for "unknown backend errors".

lzap reviewed (dismissed):
It's perfect, one nitpick about the "New" in the error message.

ezr-ondrej reviewed (approved):
Let's get it in, lets get the filtering by provider and tests in follow ups :+1:
